ksqldb-engine/src/main/java/io/confluent/ksql/utilization/PersistentQuerySaturationMetrics.java

```diff
@@ -0,0 +1,318 @@
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import org.apache.kafka.common.Metric;
+import org.apache.kafka.streams.KafkaStreams;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class PersistentQuerySaturationMetrics implements Runnable {
+  private static final Logger LOGGER
+      = LoggerFactory.getLogger(PersistentQuerySaturationMetrics.class);
+
+  private static final String QUERY_SATURATION = "query-saturation";
+  private static final String NODE_QUERY_SATURATION = "node-query-saturation";
+  private static final String QUERY_THREAD_SATURATION = "query-thread-saturation";
+  private static final String STREAMS_TOTAL_BLOCKED_TIME = "blocked-time-total";
+  private static final String STREAMS_THREAD_START_TIME = "thread-start-time";
+  private static final String STREAMS_THREAD_METRICS_GROUP = "stream-thread-metrics";
+  private static final String THREAD_ID = "thread-id";
+  private static final String QUERY_ID = "query-id";
+
+  private final Map<String, KafkaStreamsSaturation> perKafkaStreamsStats = new HashMap<>();
+  private final KsqlEngine engine;
+  private final Supplier<Instant> time;
+  private final MetricsReporter reporter;
+  private final Duration window;
+  private final Duration sampleMargin;
+
+  public PersistentQuerySaturationMetrics(
+      final KsqlEngine engine,
+      final MetricsReporter reporter,
+      final Duration window,
+      final Duration sampleMargin
+  ) {
+    this(Instant::now, engine, reporter, window, sampleMargin);
+  }
+
+  @VisibleForTesting
+  PersistentQuerySaturationMetrics(
+      final Supplier<Instant> time,
+      final KsqlEngine engine,
+      final MetricsReporter reporter,
+      final Duration window,
+      final Duration sampleMargin
+  ) {
+    this.time =  Objects.requireNonNull(time, "time");
+    this.engine = Objects.requireNonNull(engine, "engine");
+    this.reporter = Objects.requireNonNull(reporter, "reporter");
+    this.window = Objects.requireNonNull(window, "window");
+    this.sampleMargin = Objects.requireNonNull(sampleMargin, "sampleMargin");
+  }
+
+  @Override
+  public void run() {
+    final Instant now = time.get();
+
+    try {
+      final Collection<PersistentQueryMetadata> queries = engine.getPersistentQueries();
+      final Optional<Double> saturation = queries.stream()
+          .map(q -> measure(now, q))
+          .max(PersistentQuerySaturationMetrics::compareSaturation)
+          .orElse(Optional.of(0.0));
+      saturation.ifPresent(s -> report(now, s));
+
+      final Set<String> appIds = queries.stream()
+          .map(PersistentQueryMetadata::getQueryApplicationId)
```

Comment:
Can you use QueryID instead? That will work better with the shared runtimes.

Reply:
On the contrary I think it makes sense to track saturation at the runtime/app level. At the lowest level it's a property of a stream thread. We then take the max saturation over all stream threads to compute runtime-level saturation. And max over all runtimes to compute node-level saturation.

Reply:
That makes sense. It looks like the metrics are still being created for each query, can you filter out duplicates? Otherwise I think we get an illegal argument exception